pkg/controllers/dashboard/systemcharts/controller.go

```diff
@@ -232,14 +240,47 @@ func (h *handler) getChartsToInstall() []*chart.Definition {
 // when a specific event occurs on the target deployment. It is currently used to manage
 // the migration from the legacy k3s-based-upgrader app to the system-upgrade-controller app.
 func (h *handler) onDeployment(_ string, d *k8sappsv1.Deployment) (*k8sappsv1.Deployment, error) {
+	var found bool
 	if d != nil && d.Namespace == namespace.System && d.Name == sucDeploymentName {
+		appName, ok := d.Labels[legacyAppLabel]
+		if ok {
+			if appName == k3sbasedupgrade.K3sAppName || appName == k3sbasedupgrade.Rke2AppName {
+				found = true
+			}
+		}
+	}
+	if found {
+		index := slices.Index(d.Finalizers, legacyAppFinalizer)
+		logrus.Infof("[systemcharts] found deployment %s/%s with label %s=%s, index of target finalzier = %d",
+			d.Namespace, d.Name, legacyAppLabel, d.Labels[legacyAppLabel], index)
+		d = d.DeepCopy()
+		// When the deployment is being deleted, remove the finalizer if it exists, and enqueue the rancher-charts clusterRepo
 		if d.DeletionTimestamp != nil {
-			appName, ok := d.Labels[legacyAppLabel]
-			if ok {
-				logrus.Debugf("[systemcharts] found deployment %s/%s with label %s=%s", d.Namespace, d.Name, legacyAppLabel, appName)
-				if appName == k3sbasedupgrade.K3sAppName || appName == k3sbasedupgrade.Rke2AppName {
-					logrus.Debugf("[systemcharts] enqueue %s", repoName)
-					h.clusterRepo.Enqueue(repoName)
+			if index >= 0 {
+				var finalizers []string
+				finalizers = append(finalizers, d.Finalizers[:index]...)
+				finalizers = append(finalizers, d.Finalizers[index+1:]...)
+				d.Finalizers = finalizers
+				err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+					_, err := h.deployment.Update(d)
+					return err
+				})
+				if err != nil {
+					return nil, fmt.Errorf("[systemcharts] failed to update deployment %s/%s: %w", d.Namespace, d.Name, err)
+				}
+			}
+			logrus.Infof("[systemcharts] enqueue %s", repoName)
+			h.clusterRepo.EnqueueAfter(repoName, 2*time.Second)
+		} else {
+			// Add the finalizer if it is absent, ensuring Wrangler can detect the deletion event
+			if index == -1 {
+				d.Finalizers = append(d.Finalizers, legacyAppFinalizer)
+				err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+					_, err := h.deployment.Update(d)
+					return err
+				})
+				if err != nil {
+					return nil, fmt.Errorf("[systemcharts] failed to update deployment %s/%s: %w", d.Namespace, d.Name, err)
```

Comment:
```suggestion
return nil, fmt.Errorf("failed to update deployment %s/%s: %w", d.Namespace, d.Name, err)
```
component shouldn't be in error message.